### PR TITLE
Fix (almost all) issues found by dialyzer

### PR DESCRIPTION
**Pull request by paulo-ferraz-oliveira**
My initial goal was to add -export_type([arn/0]) to erlcloud_ecs.erl, because it's used by erlcloud_ec2.hrl but not visible.
I ended up dialyzing the whole thing and fixing issues as I found them.

So here's what's happened, after all those commits...

I didn't care for the undocumented use (working, though, but still...) of application:get_env/set_env/unset_env in erlcloud_aws.erl. The warnings state "breaking of contract" and "no match", and these are left unfixed as far as dialyzer is concerned.

Regarding the 'undefined' -spec(). in records (which I added quite a few of), it is basically "imposed" by OTP 19+, if I recall correctly.
I did it only to those fields that caused warnings.

Finally, I tried to break down the commits into different modules, except where this was not viable due to tight dependencies.

There is a warning still (not counting erlcloud_aws'), in erlcloud_route53.erl, that relates to the fact the function isn't exported (except for TEST) and internally Args is always a list (it's 1 in the tests suites, though).

All the changes are basically restricted to -spec().s, but I'll state below, the ones that change code too, or where more arguable decisions were made:
- include/erlcloud_xmerl.hrl was created to allow for centralized (since xmerl does not have the required types exported) -spec()ing
- erlcloud_aws's update_config/1 was updated to reflect the fact that get_role_credentials(Config) doesn't return {error, _}
- erlcloud_aws's profiles_recurse/5 was updated to reflect the fact that profiles_resolve only returns {ok, _}
- erlcloud_sns' publish_to_phone/2 seemed to erroneously call its /4 counterpart, and I fixed it
- erlcloud_ec2's ecs_request_no_update/3 never accepts <<>>, so I removed that clause
- erlcloud_emr's request_no_update/8 was updated to reflect the fact that there are no errors apart from http_error and socket_error
- erlcloud_inspector's localize_text/1 uses string instead of binary to respect internal -spec()
- erlcloud_waf's waf_request_no_update/3 never accepts <<>>, so I removed that clause
- in functions that depend (from their lowest level to the highest level) on erlang:error, I added no_return() to the -spec().s

`make check` was executed in OTP 17.5.6, 18.3, 19.1 and 20.2, for rebar3.

### CODE REVIEW

**Comment by motobob** [comment]
that is an impressive effort ! great thanks!
it will take me couple days to double check at least but we are going to merge this eventually for sure.

**Comment by paulo-ferraz-oliveira** [comment]
Yup, thanks!

It took me some hours of work and "fighting" against the dialyzer, but I got to a point I'm comfortable with the result. The one that gave me the biggest headache, simply because the dialyzer kept saying "no local return" was https://github.com/erlcloud/erlcloud/pull/474/commits/688767f3d840cb726913f208dba374afa5db14c0. I eventually followed the code execution path very closely (it's very easy to miss a [string()] when below, in the function body, you have list_to_binary(_) which I didn't know would accept [string()] so easily), and there you go!, it solved a few dozen warnings, especially from erlcloud_sdb.erl and erlcloud_aws.erl.

I'll keep an out out for any comments so I may proceed with fixes until the merge and eventual patch release.

**Comment by paulo-ferraz-oliveira** [comment]
Rebased the PR branch on top of the new (previously conflicting) changes and added a new fix for a newly found dialyzer -related issue

**Comment by nalundgaard** [comment]
Thanks for this effort, @paulo-ferraz-oliveira! This looks pretty cool. I think that I may have introduced a merge conflict in my latest PR (#479), among maybe a few others. Do you think you could rebase again?

@motobob I haven't exhaustively checked this at all, but it looks like a good improvement to many bad practices. I suppose we should prioritize this getting feedback and merged ASAP, as it's the kind of sweeping change that will be screwed up by merge conflicts almost every time we merge anyone else's PR.

**Comment by motobob** [comment]
I'm not that much of a fan for records with `field :: undefined | term()` approach. as this just make dializer silent while not fixing the problems.
Obviously, the issue here is use of records them self with too many optional fields. 

I've merge the conflict and will merge this to as you said not rebase it every time.

File: include/erlcloud_mon.hrl [hunk 1]
```diff
@@ -57,13 +57,13 @@
 %% @end
 %%------------------------------------------------------------------------------
 -record(metric_datum, {
-          metric_name       ::string(),         %% The name of the metric.
+          metric_name       ::string(),                    %% The name of the metric.
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply include/erlcloud_mon.hrl]: I don't usual change code style for projects that aren't mine, so this might have been an unintended change.

**Comment by paulo-ferraz-oliveira** [comment]
@motobob: I'm not a huge fan of `field :: undefined | term()` either, especially since going from OTP 18 to OTP 19 makes for a difficult few dialyzer passes, initially. I agree with you when you say the issue is the use of records (we could think about migrating to maps in a future version?, as they make more sense for options, than records). In any case, we are making "the dialyzer silent" _while_ making true the way we are using the records here.

**Comment by paulo-ferraz-oliveira** [comment]
> Thanks for this effort, @paulo-ferraz-oliveira! This looks pretty cool. I think that I may have introduced a merge conflict in my latest PR (#479), among maybe a few others. Do you think you could rebase again?

@nalundgaard: sorry about that, I didn't see the most recent conflicts. I did solve one, recently, but I don't get notified (at least I don't know if it's possible that GitHub does that) for merges that start causing conflicts in open PRs. It would be a neat feature, though...

**Pull request merged**